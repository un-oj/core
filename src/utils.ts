export { version } from '../package.json' with { type: 'json' };

/** Get the first key of an object. */
export function getFirstKey<T extends Record<never, never>>(obj: T): keyof T {
  return Object.keys(obj)[0] as keyof T;
}

/**
 * Parses a human-readable time string.
 * @param s The string to parse.
 * @returns The time in milliseconds, or undefined if failed.
 */
export function parseTime(s: string): number | undefined {
  const val = Number.parseInt(s);
  if (Number.isNaN(val))
    return;
  switch (s) {
    case `${val} second`:
    case `${val} seconds`:
    case `${val} sec`:
    case `${val} secs`:
      return val * 1000;
  }
}

const MEMORY_UNITS: Record<string, number> = {
  megabyte: 1024 * 1024,
  megabytes: 1024 * 1024,
  MB: 1024 * 1024,
  MiB: 1024 * 1024,
};

/**
 * Parses a human-readable memory size string.
 *
 * @param s The string to parse.
 * @returns The memory size in bytes, or undefined if failed.
 */
export function parseMemory(s: string): number | undefined {
  const [, val, unit] = s.match(/^(\d+) ([a-z]+)$/i) ?? [];
  if (!val || !unit)
    return;
  return Number(val) * MEMORY_UNITS[unit];
}

/**
 * Adds key-value pairs to the provided headers object if they aren't present.
 *
 * @param h The original headers object.
 * @param add The key-value pairs to add.
 * @returns A **new** headers object.
 */
export function addHeaders(h: HeadersInit | undefined, add: Array<[string, string]>): Record<string, string> {
  if (h instanceof Headers)
    h = h.toJSON();
  if (Array.isArray(h))
    h = Object.fromEntries(h);
  else if (h == null)
    h = {};
  else
    h = structuredClone(h);

  for (const item of add) {
    if (!(item[0] in h))
      h[item[0]] = item[1];
  }
  return h;
}

/** General error class for UnOJ. */
export class UnOJError extends Error {}
