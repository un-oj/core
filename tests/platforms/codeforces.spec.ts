import { NotFoundError } from '@un-oj/core';
import Codeforces from '@un-oj/core/platforms/codeforces';
import { UnOJError } from '@un-oj/core/utils';
import { describe, expect, it } from 'bun:test';
import { assertProblem } from './utils';

const TIMEOUT = 10 * 1000;

describe('Codeforces platform', () => {
  const cf = new Codeforces();

  it('should fetch old problem', async () => {
    assertProblem(await cf.getProblem('1A'));
  }, TIMEOUT);

  it('should fetch new problem', async () => {
    assertProblem(await cf.getProblem('2050E'));
  }, TIMEOUT);

  it('should fetch interactive problem', async () => {
    assertProblem(await cf.getProblem('1486C2'), { type: 'interactive' });
  }, TIMEOUT);

  it('should fetch gym problem', async () => {
    assertProblem(await cf.getProblem('105863A'));
  }, TIMEOUT);

  it.todo('should throw w/ not-ready problem', async () => {
    expect(cf.getProblem('105851A')).rejects.toThrow(new UnOJError('Statement not ready'));
  }, TIMEOUT);

  it('should throw NotFoundError w/ non-existent problem', async () => {
    expect(cf.getProblem('114514A')).rejects.toThrow(NotFoundError);
  }, TIMEOUT);
});
