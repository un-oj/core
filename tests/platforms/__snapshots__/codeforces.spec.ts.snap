// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Codeforces platform should fetch old problem 1`] = `
{
  "description": "<p>Theatre Square in the capital city of Berland has a rectangular shape with the size <span class="tex-span"><i>n</i> × <i>m</i></span> meters. On the occasion of the city's anniversary, a decision was taken to pave the Square with square granite flagstones. Each flagstone is of the size <span class="tex-span"><i>a</i> × <i>a</i></span>.</p><p>What is the least number of flagstones needed to pave the Square? It's allowed to cover the surface larger than the Theatre Square, but the Square has to be covered. It's not allowed to break the flagstones. The sides of flagstones should be parallel to the sides of the Square.</p>",
  "difficulty": 1000,
  "id": "1A",
  "link": "https://codeforces.com/problemset/problem/1/A",
  "memoryLimit": 268435456,
  "samples": [
    {
      "input": 
"6 6 4
"
,
      "output": 
"4
"
,
    },
  ],
  "tags": [
    "math",
  ],
  "timeLimit": 1000,
  "title": "Theatre Square",
  "type": "traditional",
}
`;

exports[`Codeforces platform should fetch new problem 1`] = `
{
  "description": "<p>You are given three strings: $$$a$$$, $$$b$$$, and $$$c$$$, consisting of lowercase Latin letters. The string $$$c$$$ was obtained in the following way:</p><ol> <li> At each step, either string $$$a$$$ or string $$$b$$$ was randomly chosen, and the first character of the chosen string was removed from it and appended to the end of string $$$c$$$, until one of the strings ran out. After that, the remaining characters of the non-empty string were added to the end of $$$c$$$. </li><li> Then, a certain number of characters in string $$$c$$$ were randomly changed. </li></ol><p>For example, from the strings $$$a=\\color{red}{\\text{abra}}$$$ and $$$b=\\color{blue}{\\text{cada}}$$$, without character replacements, the strings $$$\\color{blue}{\\text{ca}}\\color{red}{\\text{ab}}\\color{blue}{\\text{d}}\\color{red}{\\text{ra}}\\color{blue}{\\text{a}}$$$, $$$\\color{red}{\\text{abra}}\\color{blue}{\\text{cada}}$$$, $$$\\color{red}{\\text{a}}\\color{blue}{\\text{cada}}\\color{red}{\\text{bra}}$$$ could be obtained.</p><p>Find the minimum number of characters that could have been changed in string $$$c$$$.</p>",
  "difficulty": 1500,
  "id": "2050E",
  "link": "https://codeforces.com/problemset/problem/2050/E",
  "memoryLimit": 268435456,
  "samples": [
    {
      "input": "<div class="test-example-line test-example-line-even test-example-line-0">7</div><div class="test-example-line test-example-line-odd test-example-line-1">a</div><div class="test-example-line test-example-line-odd test-example-line-1">b</div><div class="test-example-line test-example-line-odd test-example-line-1">cb</div><div class="test-example-line test-example-line-even test-example-line-2">ab</div><div class="test-example-line test-example-line-even test-example-line-2">cd</div><div class="test-example-line test-example-line-even test-example-line-2">acbd</div><div class="test-example-line test-example-line-odd test-example-line-3">ab</div><div class="test-example-line test-example-line-odd test-example-line-3">ba</div><div class="test-example-line test-example-line-odd test-example-line-3">aabb</div><div class="test-example-line test-example-line-even test-example-line-4">xxx</div><div class="test-example-line test-example-line-even test-example-line-4">yyy</div><div class="test-example-line test-example-line-even test-example-line-4">xyxyxy</div><div class="test-example-line test-example-line-odd test-example-line-5">a</div><div class="test-example-line test-example-line-odd test-example-line-5">bcd</div><div class="test-example-line test-example-line-odd test-example-line-5">decf</div><div class="test-example-line test-example-line-even test-example-line-6">codes</div><div class="test-example-line test-example-line-even test-example-line-6">horse</div><div class="test-example-line test-example-line-even test-example-line-6">codeforces</div><div class="test-example-line test-example-line-odd test-example-line-7">egg</div><div class="test-example-line test-example-line-odd test-example-line-7">annie</div><div class="test-example-line test-example-line-odd test-example-line-7">egaegaeg</div>",
      "output": 
"1
0
2
0
3
2
3
"
,
    },
  ],
  "tags": [
    "dp",
    "implementation",
    "strings",
  ],
  "timeLimit": undefined,
  "title": "Three Strings",
  "type": "traditional",
}
`;

exports[`Codeforces platform should fetch gym problem 1`] = `
{
  "description": "<p>March is visiting the beautiful city of Belobog!</p><p>The city of Belobog can be modeled through a coordinate system. March is currently at the point $$$(x_1,y_1)$$$ and she would like to get to the point $$$(x_2,y_2)$$$. </p><p>Due to technological advancements, most people prefer to teleport instead of walk. March can perform the following operation any number of times: </p><ul> <li> Choose two integers $$$x$$$ and $$$y$$$. Then, she will teleport to the location that is the reflection of her current location across the point $$$(x,y)$$$. </li></ul><p>Is it possible to reach $$$(x_2,y_2)$$$ through any number of operations?</p>",
  "difficulty": undefined,
  "id": "105863A",
  "link": "https://codeforces.com/gym/105863/problem/A",
  "memoryLimit": 268435456,
  "samples": [
    {
      "input": "<div class="test-example-line test-example-line-even test-example-line-0">5</div><div class="test-example-line test-example-line-odd test-example-line-1">2 3 2 5</div><div class="test-example-line test-example-line-even test-example-line-2">-15 20 -31 24</div><div class="test-example-line test-example-line-odd test-example-line-3">24 31 93 77</div><div class="test-example-line test-example-line-even test-example-line-4">0 0 0 0</div><div class="test-example-line test-example-line-odd test-example-line-5">1000000000 999999999 -999999999 -1000000000</div>",
      "output": 
"YES
YES
NO
YES
NO
"
,
    },
  ],
  "tags": [],
  "timeLimit": 2000,
  "title": "Reflecting",
  "type": "traditional",
}
`;

exports[`Codeforces platform should throw NotFoundError w/ non-existent problem 1`] = `
{
  "description": "<p>Theatre Square in the capital city of Berland has a rectangular shape with the size <span class="tex-span"><i>n</i> × <i>m</i></span> meters. On the occasion of the city's anniversary, a decision was taken to pave the Square with square granite flagstones. Each flagstone is of the size <span class="tex-span"><i>a</i> × <i>a</i></span>.</p><p>What is the least number of flagstones needed to pave the Square? It's allowed to cover the surface larger than the Theatre Square, but the Square has to be covered. It's not allowed to break the flagstones. The sides of flagstones should be parallel to the sides of the Square.</p>",
  "difficulty": 1000,
  "id": "1A",
  "link": "https://codeforces.com/problemset/problem/1/A",
  "memoryLimit": 268435456,
  "samples": [
    {
      "input": 
"6 6 4
"
,
      "output": 
"4
"
,
    },
  ],
  "tags": [
    "math",
  ],
  "timeLimit": 1000,
  "title": "Theatre Square",
  "type": "traditional",
}
`;

exports[`Codeforces platform should throw NotFoundError w/ non-existent problem 2`] = `
{
  "description": "<p>March is visiting the beautiful city of Belobog!</p><p>The city of Belobog can be modeled through a coordinate system. March is currently at the point $$$(x_1,y_1)$$$ and she would like to get to the point $$$(x_2,y_2)$$$. </p><p>Due to technological advancements, most people prefer to teleport instead of walk. March can perform the following operation any number of times: </p><ul> <li> Choose two integers $$$x$$$ and $$$y$$$. Then, she will teleport to the location that is the reflection of her current location across the point $$$(x,y)$$$. </li></ul><p>Is it possible to reach $$$(x_2,y_2)$$$ through any number of operations?</p>",
  "difficulty": undefined,
  "id": "105863A",
  "link": "https://codeforces.com/gym/105863/problem/A",
  "memoryLimit": 268435456,
  "samples": [
    {
      "input": "<div class="test-example-line test-example-line-even test-example-line-0">5</div><div class="test-example-line test-example-line-odd test-example-line-1">2 3 2 5</div><div class="test-example-line test-example-line-even test-example-line-2">-15 20 -31 24</div><div class="test-example-line test-example-line-odd test-example-line-3">24 31 93 77</div><div class="test-example-line test-example-line-even test-example-line-4">0 0 0 0</div><div class="test-example-line test-example-line-odd test-example-line-5">1000000000 999999999 -999999999 -1000000000</div>",
      "output": 
"YES
YES
NO
YES
NO
"
,
    },
  ],
  "tags": [],
  "timeLimit": 2000,
  "title": "Reflecting",
  "type": "traditional",
}
`;

exports[`Codeforces platform should throw NotFoundError w/ non-existent problem 3`] = `
{
  "description": "<p>You are given three strings: $$$a$$$, $$$b$$$, and $$$c$$$, consisting of lowercase Latin letters. The string $$$c$$$ was obtained in the following way:</p><ol> <li> At each step, either string $$$a$$$ or string $$$b$$$ was randomly chosen, and the first character of the chosen string was removed from it and appended to the end of string $$$c$$$, until one of the strings ran out. After that, the remaining characters of the non-empty string were added to the end of $$$c$$$. </li><li> Then, a certain number of characters in string $$$c$$$ were randomly changed. </li></ol><p>For example, from the strings $$$a=\\color{red}{\\text{abra}}$$$ and $$$b=\\color{blue}{\\text{cada}}$$$, without character replacements, the strings $$$\\color{blue}{\\text{ca}}\\color{red}{\\text{ab}}\\color{blue}{\\text{d}}\\color{red}{\\text{ra}}\\color{blue}{\\text{a}}$$$, $$$\\color{red}{\\text{abra}}\\color{blue}{\\text{cada}}$$$, $$$\\color{red}{\\text{a}}\\color{blue}{\\text{cada}}\\color{red}{\\text{bra}}$$$ could be obtained.</p><p>Find the minimum number of characters that could have been changed in string $$$c$$$.</p>",
  "difficulty": 1500,
  "id": "2050E",
  "link": "https://codeforces.com/problemset/problem/2050/E",
  "memoryLimit": 268435456,
  "samples": [
    {
      "input": "<div class="test-example-line test-example-line-even test-example-line-0">7</div><div class="test-example-line test-example-line-odd test-example-line-1">a</div><div class="test-example-line test-example-line-odd test-example-line-1">b</div><div class="test-example-line test-example-line-odd test-example-line-1">cb</div><div class="test-example-line test-example-line-even test-example-line-2">ab</div><div class="test-example-line test-example-line-even test-example-line-2">cd</div><div class="test-example-line test-example-line-even test-example-line-2">acbd</div><div class="test-example-line test-example-line-odd test-example-line-3">ab</div><div class="test-example-line test-example-line-odd test-example-line-3">ba</div><div class="test-example-line test-example-line-odd test-example-line-3">aabb</div><div class="test-example-line test-example-line-even test-example-line-4">xxx</div><div class="test-example-line test-example-line-even test-example-line-4">yyy</div><div class="test-example-line test-example-line-even test-example-line-4">xyxyxy</div><div class="test-example-line test-example-line-odd test-example-line-5">a</div><div class="test-example-line test-example-line-odd test-example-line-5">bcd</div><div class="test-example-line test-example-line-odd test-example-line-5">decf</div><div class="test-example-line test-example-line-even test-example-line-6">codes</div><div class="test-example-line test-example-line-even test-example-line-6">horse</div><div class="test-example-line test-example-line-even test-example-line-6">codeforces</div><div class="test-example-line test-example-line-odd test-example-line-7">egg</div><div class="test-example-line test-example-line-odd test-example-line-7">annie</div><div class="test-example-line test-example-line-odd test-example-line-7">egaegaeg</div>",
      "output": 
"1
0
2
0
3
2
3
"
,
    },
  ],
  "tags": [
    "dp",
    "implementation",
    "strings",
  ],
  "timeLimit": undefined,
  "title": "Three Strings",
  "type": "traditional",
}
`;

exports[`Codeforces platform should fetch interactive problem 1`] = `
{
  "description": "<p><span class="tex-font-style-bf">The only difference between the easy and the hard version is the limit to the number of queries</span>.</p><p><span class="tex-font-style-bf">This is an interactive problem.</span></p><p>There is an array $$$a$$$ of $$$n$$$ <span class="tex-font-style-bf">different</span> numbers. In one query you can ask the position of the second maximum element in a subsegment $$$a[l..r]$$$. Find the position of the maximum element in the array in no more than <span class="tex-font-style-bf">20</span> queries.</p><p>A subsegment $$$a[l..r]$$$ is all the elements $$$a_l, a_{l + 1}, ..., a_r$$$. After asking this subsegment you will be given the position of the second maximum from this subsegment <span class="tex-font-style-bf">in the whole</span> array.</p>",
  "difficulty": 1900,
  "id": "1486C2",
  "link": "https://codeforces.com/problemset/problem/1486/C2",
  "memoryLimit": 268435456,
  "samples": [
    {
      "input": 
"5

3

4

"
,
      "output": 
"? 1 5

? 4 5

! 1"
,
    },
  ],
  "tags": [
    "binary search",
    "interactive",
  ],
  "timeLimit": 1000,
  "title": "Guessing the Greatest (hard version)",
  "type": "interactive",
}
`;

exports[`Codeforces platform should fetch problem 2`] = `
{
  "description": "<p>You are given three strings: $$$a$$$, $$$b$$$, and $$$c$$$, consisting of lowercase Latin letters. The string $$$c$$$ was obtained in the following way:</p><ol> <li> At each step, either string $$$a$$$ or string $$$b$$$ was randomly chosen, and the first character of the chosen string was removed from it and appended to the end of string $$$c$$$, until one of the strings ran out. After that, the remaining characters of the non-empty string were added to the end of $$$c$$$. </li><li> Then, a certain number of characters in string $$$c$$$ were randomly changed. </li></ol><p>For example, from the strings $$$a=\\color{red}{\\text{abra}}$$$ and $$$b=\\color{blue}{\\text{cada}}$$$, without character replacements, the strings $$$\\color{blue}{\\text{ca}}\\color{red}{\\text{ab}}\\color{blue}{\\text{d}}\\color{red}{\\text{ra}}\\color{blue}{\\text{a}}$$$, $$$\\color{red}{\\text{abra}}\\color{blue}{\\text{cada}}$$$, $$$\\color{red}{\\text{a}}\\color{blue}{\\text{cada}}\\color{red}{\\text{bra}}$$$ could be obtained.</p><p>Find the minimum number of characters that could have been changed in string $$$c$$$.</p>",
  "difficulty": 1500,
  "id": "2050E",
  "link": "https://codeforces.com/problemset/problem/2050/E",
  "memoryLimit": 268435456,
  "samples": [
    {
      "input": "<div class="test-example-line test-example-line-even test-example-line-0">7</div><div class="test-example-line test-example-line-odd test-example-line-1">a</div><div class="test-example-line test-example-line-odd test-example-line-1">b</div><div class="test-example-line test-example-line-odd test-example-line-1">cb</div><div class="test-example-line test-example-line-even test-example-line-2">ab</div><div class="test-example-line test-example-line-even test-example-line-2">cd</div><div class="test-example-line test-example-line-even test-example-line-2">acbd</div><div class="test-example-line test-example-line-odd test-example-line-3">ab</div><div class="test-example-line test-example-line-odd test-example-line-3">ba</div><div class="test-example-line test-example-line-odd test-example-line-3">aabb</div><div class="test-example-line test-example-line-even test-example-line-4">xxx</div><div class="test-example-line test-example-line-even test-example-line-4">yyy</div><div class="test-example-line test-example-line-even test-example-line-4">xyxyxy</div><div class="test-example-line test-example-line-odd test-example-line-5">a</div><div class="test-example-line test-example-line-odd test-example-line-5">bcd</div><div class="test-example-line test-example-line-odd test-example-line-5">decf</div><div class="test-example-line test-example-line-even test-example-line-6">codes</div><div class="test-example-line test-example-line-even test-example-line-6">horse</div><div class="test-example-line test-example-line-even test-example-line-6">codeforces</div><div class="test-example-line test-example-line-odd test-example-line-7">egg</div><div class="test-example-line test-example-line-odd test-example-line-7">annie</div><div class="test-example-line test-example-line-odd test-example-line-7">egaegaeg</div>",
      "output": 
"1
0
2
0
3
2
3
"
,
    },
  ],
  "tags": [
    "dp",
    "implementation",
    "strings",
  ],
  "timeLimit": undefined,
  "title": "Three Strings",
  "type": "traditional",
}
`;
